src/software/ai/evaluation/robot.h

```diff
@@ -28,26 +28,45 @@ bool robotOrientationWithinAngleThresholdOfTarget(Point position, Angle orientat
  * Determines if a robot has possession of the ball. A robot is considered to have
  * possession if the ball is in a area close to its dribbler.
  *
- * @param  The ball that is wanted to be possessed
- * @param robot The Robot which wants to know if it has the ball.
+ * @param ball_states The circular buffer of TimestampedBallState
+ * @param robot_states The circular buffer of TimestampedRobotState
  * @param timestamp The time at which we want to know if the robot had the ball
  * @return True if the ball is close to the front dribbler, False if the ball is not,
  *          and nullopt if we don't have information for the given timestamp. This
  * function is guaranteed to return non-nullopt if no timestamp is passed in.
  */
-std::optional<bool> robotHasPossession(const Ball& ball, const Robot& robot,
-                                       std::optional<Timestamp> timestamp = std::nullopt);
+std::optional<bool> robotHasPossession(
+    const boost::circular_buffer<TimestampedBallState>& ball_states,
+    const boost::circular_buffer<TimestampedRobotState>& robot_states,
+    std::optional<Timestamp> timestamp = std::nullopt);
 
 /**
  * Returns true if the robot is being passed to. A robot is considered to be being
  * passed to if the ball is heading toward it above a certain speed.
- * @param world The world
+ * @param ball_states The circular buffer of TimestampedBallState
+ * @param robot_states The circular buffer of TimestampedRobotState
  * @param robot The robot
  * @param timestamp the timestamp of the world and robot states we want to consider
  * @return True if the robot is being passed to at the given time, False if the robot
  * is not being passed to, and nullopt if we don't have information for the given
  * timestamp. This function is guaranteed to return non-nullopt if no timestamp is
  * passed in.
  */
-std::optional<bool> robotBeingPassedTo(const World& world, const Robot& robot,
-                                       std::optional<Timestamp> timestamp = std::nullopt);
+std::optional<bool> robotBeingPassedTo(
+    const boost::circular_buffer<TimestampedBallState>& ball_states,
+    const boost::circular_buffer<TimestampedRobotState>& robot_states,
+    std::optional<Timestamp> timestamp = std::nullopt);
+
+/**
+ * Finds an update timestamp that is close to the provided timestamp and returns the
+ * STATE corresponding to that timestamp in the history buffer.
+ *
+ * @param timestamp timestamp of the update state index we wish to fetch
+ * @param states circular buffer of STATE to scan
+ *
+ * @return STATE corresponding to timestamp or std::nullopt if no ball state found
+ * matching timestamp.
+ */
+template <typename STATE>
+std::optional<STATE> findState(boost::circular_buffer<STATE> states,
+                               Timestamp& timestamp);
```

Comment:
should this live in `world/state.h`?

Reply:
it's only being used here in eval/robot.{h,cpp}, so I think it's best to limit the scope to here. Eventually, this will be moved to software/sensorfusion/[somewhere]